src/device/pf_pdport.c

```diff
@@ -787,48 +809,120 @@ static void pf_pdport_run_peer_check (pnet_t * net, int loc_port_num)
+/**
+ * @internal
+ * Handle that an Ethernet link goes down.
+ *
+ * Trigger a diagnosis if a check is enabled.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param loc_port_num     In:    Local port number.
+ *                                Valid range:
+ *                                1 .. PNET_NUMBER_OF_PHYSICAL_PORTS.
+ */
+static void pf_pdport_handle_link_down (pnet_t * net, int loc_port_num)
 {
-   pnet_diag_source_t diag_source = {0};
-
-   pf_pdport_init_diag_source (&diag_source, loc_port_num);
+   LOG_INFO (
+      PNET_LOG,
+      "PDPORT(%d): Link went down on port %u.\n",
+      __LINE__,
+      loc_port_num);
+   pf_lldp_stop_peer_timeout (net, loc_port_num);
+   pf_lldp_invalidate_peer_info (net, loc_port_num);
+   pf_pdport_peer_indication (net, loc_port_num);
+}
 
-   LOG_DEBUG (
+/**
+ * @internal
+ * Handle that an Ethernet link goes up.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param loc_port_num     In:    Local port number.
+ *                                Valid range:
+ *                                1 .. PNET_NUMBER_OF_PHYSICAL_PORTS.
+ */
+static void pf_pdport_handle_link_up (pnet_t * net, int loc_port_num)
+{
+   LOG_INFO (
       PNET_LOG,
-      "PDPORT(%d): Sending no-peer-detected alarm for port %u",
+      "PDPORT(%d): Link went up on port %u.\n",
       __LINE__,
       loc_port_num);
+}
 
-   (void)pf_diag_std_add (
-      net,
-      &diag_source,
-      PNET_DIAG_CH_PROP_TYPE_UNSPECIFIED,
-      PNET_DIAG_CH_PROP_MAINT_FAULT,
-      PF_WRT_ERROR_REMOTE_MISMATCH,
-      PF_WRT_ERROR_NO_PEER_DETECTED,
-      0,
-      0);
+/**
+ * @internal
+ * Monitor the Ethernet link status
+ *
+ * Set approprite diagnosis, if check is enabled.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param loc_port_num     In:    Local port number.
+ *                                Valid range:
+ *                                1 .. PNET_NUMBER_OF_PHYSICAL_PORTS.
+ */
+static void pf_pdport_monitor_link (pnet_t * net, int loc_port_num)
+{
+   pnal_eth_status_t status;
+   pf_port_t * p_port_data = pf_port_get_state (net, loc_port_num);
+
+   if (pnal_eth_get_status (p_port_data->netif.name, &status) != 0)
+   {
+      LOG_ERROR (
+         PNET_LOG,
+         "PDPORT(%d): Failed to read link status for port %u.\n",
+         __LINE__,
+         loc_port_num);
+      return;
+   }
+
+   if (p_port_data->netif.previous_is_link_up == false && status.running)
+   {
+      pf_pdport_handle_link_up (net, loc_port_num);
+   }
+   else if (p_port_data->netif.previous_is_link_up && status.running == false)
+   {
+      pf_pdport_handle_link_down (net, loc_port_num);
+   }
+
+   p_port_data->netif.previous_is_link_up = status.running;
 }
 
 void pf_pdport_periodic (pnet_t * net)
 {
    int port;
    pf_port_iterator_t port_iterator;
    pf_port_t * p_port_data = NULL;
+   static int counter = 0;
+
+   // TODO Move somewhere else
```

Comment:
idea:
Move monitor functionality into separate function which is registered to scheduler, run every 100ms. 

Maybe we should measure the time it takes to make the system call. If needed, only poll status of one network interface at a time. The the port_iterator variable can be stored in port data and only one port is checked every iteration.

On the other hand, if system call is fast we move pf_pdport_monitor_link (net, port); into 
existing iterator loop which would be simplest solution.

Reply:
Implemented a scheduled function that scan the Ethernet links